### PR TITLE
Fix decimals float on number field #2372

### LINKED ISSUE #2372: Number field cuts of decimal places
To my surprise the number field doesn't work any more if more then 4 decimal places are used.

**Working:**

```
                lat:
                    label:        Breitengrad
                    type:         number
                    step:         0.0001
```

**Not working anymore:**

```
                lat:
                    label:        Breitengrad
                    type:         number
                    step:         0.00001
```

If you try with 5 or more decimal places, all decimal places are silently dropped (you loose everything following the comma, making the field an integer-only field).

If this is expected, it doesn't make sense to me, because decimal numbers are not limited to 4 decimal places. In some situations - like with latitudes or longitudes - we need more then 4 decimal places in the numbers. What is the best way to store such numbers, if not in the numbers field?

### PR DESCRIPTION
**Pull request by afbora**
## Describe the PR

This issue was about casting specific float number types (like 1.0E-6) to string. The problem was difficult to find, but easy to solve 😅 

## Related issues

- Fixes #2372 

## Ready?

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Fixed code style issues with CS fixer and `composer fix`
- [x] Added in-code documentation (if needed)

### CODE REVIEW

**Review by lukasbestle — COMMENTED** [review]
This doesn't fix the original issue for me – the float is later converted to string, which will again result in the exponential syntax that won't be understood by the browser.

**Comment by afbora** [comment]
@lukasbestle Original issue? I tested on panel with `step: 0.00001` option before create PR and worked properly. I have tested now and step option works great 🤔

**Comment by afbora** [comment]
Btw scientific notation is supported all major browsers

https://caniuse.com/#feat=mdn-css_types_number_scientific_notation

**Comment by afbora** [comment]
As I mentioned in my first message, the source of the error was due to the insertion of the scientific notation type into the `str_replace()` method in `Str::float()` method.

**Review by lukasbestle — DISMISSED** [review]
You are right, it *does* fix the original issue. Seems like I tested with a few too many zeroes last time, because `0.0000001` still results in the following error in the Panel:

```
Cannot read property '$invalid' of null
```

But that seems to be unrelated to this issue as the API response for such a `step` definition is correct and the error occurs only on the frontend.

**Comment by afbora** [comment]
Yes, it gives this error in much smaller numbers. Actually similar errors. The error we solved was related to the backend, this error is on the javasript side. It seems like I have solved the error on the javascript side, should we continue under this PR?

Issue about javascript side about casting string:
https://github.com/getkirby/kirby/blob/3.3.3/panel/src/components/Forms/Input/NumberInput.vue#L87

**Comment by lukasbestle** [comment]
Oh yes, that's exactly the same issue in the frontend.

> It seems like I have solved the error on the javascript side, should we continue under this PR?

If you have an easy fix for this: Sure!

**Comment by afbora** [comment]
@lukasbestle I have fixed the issue on javascript side. It works successfully in my tests.

@distantnative @bastianallgeier There may be more descriptive method and variable names 😇  I may have made incorrect or incomplete use in `number`, `value`, and `step` assignments. I would appreciate if you fix it or make improvements. Thank you 🙏 

**Reference solution**
http://blog.davidjs.com/2018/07/convert-exponential-numbers-to-decimal-in-javascript/

**Comment by bastianallgeier** [comment]
I can't believe how much pain a simple number field already cost us. The convert method seems quite extensive :( I wish there was a way to solve this with less code.

**Comment by afbora** [comment]
@bastianallgeier What do you think now?

**Comment by afbora** [comment]
I'll do some improvements

**Comment by afbora** [comment]
@bastianallgeier Ready to review 👍

File: panel/src/components/Forms/Input/NumberInput.vue [hunk 1]
```diff
@@ -84,6 +85,10 @@ export default {
         return 0;
       }
 
+      if (step.toString().indexOf('e') !== -1) {
+        return parseInt(step.toFixed(16).split(".")[1].split("").reverse().join("")).toString().length;
+      }
```
Comment (afbora) [inline panel/src/components/Forms/Input/NumberInput.vue:90]: Steps of method that i found:

1. Check float: 1e-6
2. Fix with 16 decimals as `0.0000010000000000`
3. Get only decimals as `0000010000000000`
4. Reverse the number: `0000000000100000`
5. ParseInt as `100000`
6. Get length as `6`

Wohooo 😄

**Review by bastianallgeier — APPROVED** [review]

**Pull request merged**